extern/test_bmi_cpp/src/test_bmi_cpp.cpp

```diff
@@ -0,0 +1,429 @@
+#include "test_bmi_cpp.hpp"
+
+#include <algorithm>
+#include <cassert>
+#include <cstring>
+
+std::string TestBmiCpp::GetComponentName(){
+  return "Testing BMI C++ Model";
+}
+
+double TestBmiCpp::GetCurrentTime(){
+  return current_model_time;
+}
+
+double TestBmiCpp::GetEndTime(){
+  double time = this->GetStartTime();
+  time += this->num_time_steps * this->time_step_size;
+  return time;
+}
+
+int TestBmiCpp::GetGridRank(const int grid){
+  if (grid == 0) {
+      return 1;
+  }
+  else {
+    throw new std::runtime_error("Rank requested for non-existent grid.");
+  }
+}
+
+int TestBmiCpp::GetGridSize(const int grid){
+  if (grid == 0) {
+      return 1;
+  }
+  else {
+    throw new std::runtime_error("Size requested for non-existent grid.");
+  }
+}
+
+std::string TestBmiCpp::GetGridType(const int grid){
+  if (grid == 0) {
+      return "scalar";
+  }
+  else {
+    throw new std::runtime_error("Type requested for non-existent grid.");
+  }
+}
+
+
+std::vector<std::string> TestBmiCpp::GetInputVarNames(){
+  return input_var_names;  
+}
+std::vector<std::string> TestBmiCpp::GetOutputVarNames(){
+  return output_var_names;
+}
+int TestBmiCpp::GetInputItemCount(){
+  return input_var_names.size();
+}
+int TestBmiCpp::GetOutputItemCount(){
+  return output_var_names.size();
+}
+
+double TestBmiCpp::GetStartTime(){
+  return 0.0;
+}
+
+double TestBmiCpp::GetTimeStep(){
+  return this->time_step_size;
+}
+
+std::string TestBmiCpp::GetTimeUnits(){
+  return "s";
+}
+
+void TestBmiCpp::GetValue(std::string name, void* dest){
+  int inds[] = {0};
+  this->GetValueAtIndices(name, dest, inds, 1);
+}
+
+void TestBmiCpp::GetValueAtIndices(std::string name, void* dest, int* inds, int len){
```

Comment:
@mattw-nws is there a particular reason the param is named `len` here when elsewhere (including in the exception message string below) it is `count`?

Reply:
Aha, you fell into my trap!

https://github.com/NOAA-OWP/ngen/blame/d11eba448683cb5d1dbdab4a6b8830b61d39d915/extern/test_bmi_c/src/bmi_test_bmi_c.c#L281

Or you're trolling me.

Seriously, yes, this should probably be `count`, and `bmi.hpp` names it such...I wonder if a previous version of `bmi.hpp` used `len`, since this parameter's existence is not strictly part of the API signature but is language-dependent (auto-generated?). 

I thought preserving parity with the existing C test code would make review easier for everyone but that's proving to be the opposite of true. :-D

Reply:
Resolved in [4ae1801](https://github.com/NOAA-OWP/ngen/pull/343/commits/4ae18013029efae6f5a4a6d948fb0b1f7245a8ed)